modules/videopress/utility-functions.php

```diff
@@ -87,13 +85,21 @@ function videopress_get_attachment_id_by_url( $url ) {
 	// Is URL in uploads directory?
 	if ( false !== strpos( $url, $dir ) ) {
 
-		$file = basename( $url );
+		$file        = basename( $url );
+		$cache_key   = 'videopress_get_attachment_id_by_url_' . md5( $url );
+		$cache_group = 'videopress';
+		$cached_id   = wp_cache_get( $cache_key, $cache_group );
```

Comment:
Quick question; did you consider using transients instead of object caching? Would it be better for performance on sites that do not use caching plugins?

Reply:
Good question, I'll look at that.

Reply:
Sure, I could add a transient here if you'd like.

But since the attachment ID is from a  database query (WP_Query), I'm not sure that using a transient to store it in the database (wp_options) would be the best approach. 

It would probably improve performance for sites without persistent object caching. But it could fill the `wp_options` table.

Reply:
@kienstra Also, is `videopress_get_attachment_id_by_url` ever called?

I _think_ the original usage was in `videopress_shortcode_override_for_core_shortcode` (771097dc2b1943a3095a9a34c586df85b4816a12), but has been superseded by 32dbcd7ac9ace94ce78b69831b2e35e7d55e9fa2, but the method itself has never been removed.

I ran `git log -p -S videopress_get_attachment_id_by_url` to confirm.

Reply:
Uf, you're right. I missed that it's not called in Jetpack. 

The diff for that is removed in https://github.com/Automattic/jetpack/pull/14803/commits/96a565aa5734c374087016b2aec56f94d102b6e4.

Reply:
Hi @jeherve, hope you had a great day.

Is there possibly another codebase where `videopress_get_attachment_id_by_url()` is called? 

If not, I could remove it here if you'd like. Though maybe that could be considered in a separate issue or PR.

Reply:
It isn't used on WordPress.com right now, so it may be safe to remove. Some third party plugin / theme authors may rely on it though, so it may be best to keep it around for now, but with a `_deprecated_function` notice so we can completely remove it in a few releases.

Reply:
Good idea, it's deprecated in https://github.com/Automattic/jetpack/pull/14803/commits/67df908fde1033325537cb41571048c7a4d92b4f

Reply:
I checked the VIP Go GH organization too -- no usage. That's the other big place I would expect something random to use it.

Reply:
The discussion here on transients as a fallback for sites without caching is still relevant for `video_get_post_by_guid()`.

>@kienstra Expiring transients aren't autoloaded and are deleted once they expire. Do you think we need to worry about polluting wp_options under these circumstances?

It's probably not likely to fill the `wp_options` table, just on sites with a huge amount of videos and traffic.

But it seems that storing the result a database query (WP_Query) in the database isn't the best approach.

I couldn't find an example in Jetpack of storing a `WP_Query` result in a transient, though there could be examples. There are a few [non-cached $wpdb queries](https://github.com/Automattic/jetpack/blob/35c35c5a2b13446e68489210ee3edc5b12c8ed73/class.jetpack.php#L3588) stored in transients.

Still, if you have strong feelings on this, I could add transient usage. There probably isn't much risk.

Reply:
Hi everyone! @jeherve asked me to share some thoughts on this discussion. 

I agree that we could use transients here. It would be more effective, since it would last through multiple requests.

Transients are not necessarily stored on the database. If memcached is enabled, for example, this option is used. And the downsize is that the expiration time is actually the maximum expiration time. So the transient may disappear earlier and we don't have much control over how much it will last. But I think this is ok for us here.

BUT, I do agree with @kienstra and I don't feel super comfortable storing the whole object in the transient. It might be harmless but if we have an edge case we are not considering here, this would make things worse.

If we would go that way, I'd say to add an extra check to see if we have a valid object before using it, just in case it got corrupted somehow.

I think a good approach here can be:

1. Use transients, but store only the post ID
2. We could use the object cache to store the post object, avoiding multiple calls to `get_post()` if we want.

And to wrap things up, I'd get the post ID in a separate method, which would be a direct replacement for the method we are deprecating :) .

Reply:
>1. Use transients, but store only the post ID
>2. We could use the object cache to store the post object, avoiding multiple calls to get_post() if we want.

Thanks, what do you think about the [latest commits](https://github.com/Automattic/jetpack/compare/67df908fde1033325537cb41571048c7a4d92b4f...21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5), which apply those 2 points?

Reply:
>And to wrap things up, I'd get the post ID in a separate method, which would be a direct replacement for the method we are deprecating :) .

Hm, did you have in mind something like:

```php
/**
 * Using a GUID, find a post ID.
 *
 * @param string $guid The guid to use to find the post ID.
 * @return int|false The post ID.
 */
function video_get_post_id_by_guid( $guid ) {
	$post = video_get_post_by_guid( $guid );

	if ( is_object( $post ) && 'WP_Post' === get_class( $post ) ) {
		return $post->ID;
	}

	return false;
}
```

Reply:
> Thanks, what do you think about the [latest commits](https://github.com/Automattic/jetpack/compare/67df908fde1033325537cb41571048c7a4d92b4f...21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5), which apply those 2 points?

What I thought is that we can use both transients and wp_cache at the same time, each one for a purpose. Transients to store the post ID, so we avoid making the costly SQL SELECT many times though several requests. Object Caching to avoid calling `get_post()` several times within the same request.

> Hm, did you have in mind something like:

Not exactly. I would actually leave the `WP_Query` call in `video_get_post_id_by_guid`, adding the `'fields' => 'ids'`. I would call it from `video_get_post_by_guid` and then call `get_post`. So I would have the same behavior every time. (Always fetch the ID and then call `get_post`)

Looks a bit cleaner to me. Each method doing just one thing, and avoiding things like `$cached_post` sometimes holding the actual post, and sometimes holding just the post id.

Also it's important to add an expiration time for the transient.

This is how I see it. (Please consider this just as an opinion!)

```PHP

function video_get_post_by_guid( $guid ) {
	// ...
	$cached_post = wp_cache_get( $cache_key, $cache_group );

	if ( $cached_post ) {
		return $cached_post;
	}

	$post_id = $this->video_get_post_id_by_guid( $guid );

	if ( is_int( $post_id ) ) {
		$post = get_post( $cached_post );
		wp_cache_set( $cache_key, $post, $cache_group, HOUR_IN_SECONDS );
		if ( $post ) {
			return $post;
		}
	}

}

function video_get_post_id_by_guid( $guid ) {
	// ... 
	$cached_id = get_transient( $cache_key );
	if ( $cached_id ) {
		return $cached_id;
	}

	// do the query
	set_transient( $cache_key, $id, HOUR_IN_SECONDS );
	return $id;
}
```

Reply:
Thanks, and sorry for the delay. I'm working on this.

Reply:
How does [this look](https://github.com/Automattic/jetpack/compare/21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5...331f265212a39d91d1447f4d316ed6374fc5b3b6)? It mainly uses your [snippet above](https://github.com/Automattic/jetpack/pull/14803#discussion_r393672549).